### PR TITLE
Truncate long tag and param values to maximum length

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11208?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11208/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11208
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Truncate long tag and param values to maximum length.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/validation.py [hunk 1]
```diff
@@ -321,10 +283,6 @@ def _validate_batch_log_data(metrics, params, tags):
         # changes to other APIs. See related discussion in
         # https://github.com/mlflow/mlflow/issues/985
         _validate_length_limit("Metric name", MAX_ENTITY_KEY_LENGTH, metric.key)
-    for param in params:
-        _validate_param(param.key, param.value)
-    for tag in tags:
-        _validate_tag(tag.key, tag.value)
```
Comment (harupy) [inline mlflow/utils/validation.py:327]: `params` and `tags` should be already validated at this point.

File: mlflow/environment_variables.py [hunk 2]
```diff
@@ -511,3 +511,11 @@ def get(self):
...
     "MLFLOW_GATEWAY_RATE_LIMITS_STORAGE_URI", str, None
 )
+
+#: If set to True, the following entities will be truncated to their maximum length:
+#: - Param value
+#: - Tag value
+#: If set to False, an exception will be raised if the length of the entity exceeds the maximum
+#: length.
+#: (default: ``True``)
+MLFLOW_TRUNCATE_LONG_ENTITIES = _BooleanEnvironmentVariable("MLFLOW_TRUNCATE_LONG_ENTITIES", True)
```
Comment (harupy) [inline mlflow/environment_variables.py:521]: A safe flag for users who don't like auto-truncation.

File: mlflow/entities/run_tag.py [hunk 3]
```diff
@@ -1,14 +1,51 @@
...
+    def validate_value(value):
+        length = len(value)
+        if length <= MAX_TAG_VAL_LENGTH:
+            return value
+
+        if MLFLOW_TRUNCATE_LONG_ENTITIES.get():
+            _logger.warning(
+                f"Tag value {value[:250]!r} ({length} characters) is truncated to "
+                f"{MAX_TAG_VAL_LENGTH} characters to meet the length limit."
+            )
```
Comment (harupy) [inline mlflow/entities/run_tag.py:41]:
```suggestion
            _logger.warning(
                f"Tag value {value[:250]!r} ({length} characters) is truncated to "
                f"{MAX_TAG_VAL_LENGTH} characters to meet the length limit. To turn this off, blah blah blah"
            )
```

We could include how to turn off auto-truncation in the message, but that would be annoying if you have multiple long values.

File: mlflow/entities/param.py [hunk 4]
```diff
@@ -19,6 +34,48 @@ def __init__(self, key, value):
...
+        if not _VALID_PARAM_AND_METRIC_NAMES.match(key):
+            raise MlflowException.invalid_parameter_value(
+                f"Invalid parameter key: '{key}'. {_BAD_CHARACTERS_MESSAGE}",
+            )
+        if path_not_unique(key):
+            raise MlflowException.invalid_parameter_value(
+                f"Invalid parameter key: '{key}'. {bad_path_message(key)}",
+            )
+
+        _validate_length_limit("Param key", MAX_ENTITY_KEY_LENGTH, key)
```
Comment (WeichenXu123) [inline mlflow/entities/param.py:58]: Q: why not define another constant like `MAX_PARAM_KEY_LENGTH` ?

File: mlflow/entities/param.py [hunk 5]
```diff
@@ -19,6 +34,48 @@ def __init__(self, key, value):
...
+
+    @staticmethod
+    def validate_value(value):
+        length = len(value)
+        if length <= MAX_PARAM_VAL_LENGTH:
+            return value
+
+        if MLFLOW_TRUNCATE_LONG_ENTITIES.get():
+            _logger.warning(
+                f"Param value {value[:250]!r} ({length} characters) is truncated to "
```
Comment (WeichenXu123) [inline mlflow/entities/param.py:69]:
```suggestion
                f"Param value {value[:250] + '...'!r} ({length} characters) is truncated to "
```

File: mlflow/environment_variables.py (hunk 2 above)
Comment (WeichenXu123) [inline mlflow/environment_variables.py:521]: shall we name it as "MLFLOW_TRUNCATE_LONG_ENTITIE_VALUE" ?

File: mlflow/utils/validation.py [hunk 6]
```diff
@@ -270,13 +278,25 @@ def _validate_tag_name(name):
...
+
+    if truncate and MLFLOW_TRUNCATE_LONG_VALUES.get():
+        _logger.warning(
+            f"{entity_name} {value[:100]}...' ({len(value)} characters) is truncated to "
+            f"{limit} characters to meet the length limit."
         )
+        return value[:limit]
+
+    raise MlflowException(
+        f"{entity_name} '{value[:250]}' had length {len(value)}, "
```
Comment (WeichenXu123) [inline mlflow/utils/validation.py:296]:
```suggestion
        f"{entity_name} '{value[:250]...}' had length {len(value)}, "
```

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: mlflow/utils/validation.py (hunk 6 above)
Comment (harupy) [reply mlflow/utils/validation.py:296]: I didn't want to break existing tests.

**Pull request merged**